zio/bzngio/index.go

```diff
@@ -0,0 +1,175 @@
+package bzngio
+
+import (
+	"errors"
+	"io"
+	"os"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+type Index struct {
+	index      []mark
+	indexReady bool
+}
+
+type IndexReader interface {
+	zbuf.ReadCloser
+	// count of how many records were read from disk, currently just
+	// used for testing.
+	Reads() uint64
+}
+
+type mark struct {
+	Ts     nano.Ts
+	Offset int64
+}
+
+// NewIndex creates a new Index object, which is the container that holds
+// the in-memory index for a (b)zng file.  The first call to NewReader()
+// will return a reader that scans the entire file, building a time-based
+// index in the process, subsequent readers can use this index to read
+// only the relevant zng streams from the underlying file.
+func NewIndex() Index {
+	return Index{}
+}
+
+func (i *Index) NewReader(f *os.File, zctx *resolver.Context, span nano.Span) (IndexReader, error) {
+	if i.indexReady {
+		return newRangeReader(f, zctx, i.index, span)
+	} else {
+		return &indexReader{
+			Reader: *NewReader(f, zctx),
+			Closer: f,
+			start:  span.Ts,
+			end:    span.End(),
+			parent: i,
+		}, nil
+	}
+}
+
+// indexReader is a zbuf.Reader that also builds an index as it reads.
+type indexReader struct {
+	Reader
+	io.Closer
+	start   nano.Ts
+	end     nano.Ts
+	parent  *Index
+	marks   []mark
+	lastSOS int64
+	lastTs  nano.Ts
+}
+
+func (i *indexReader) Read() (*zng.Record, error) {
+	for {
+		rec, err := i.readOne()
+		if err != nil {
+			return nil, err
+		}
+
+		if rec == nil {
+			i.parent.index = i.marks
+			i.parent.indexReady = true
+			return nil, nil
+		}
+
+		if rec.Ts < i.start {
+			continue
+		}
+		if rec.Ts <= i.end {
+			return rec, nil
+		}
+
+		// This record falls after the end of the requested time
+		// span.  Spin through this loop until we hit EOF anyway
+		// to finish building the index.
+		// XXX this will be wasteful if small ranges near the
+		// start of the file is all that is ever read.  revisit this...
```

Comment:
(fwiw, the current seems just right to me for a volatile index. if/when we persist it, then the index could be created at ingest time, where we read everything anway...)